File: ruby/practice/classic_puzzle/easy/organic-compound.rb
# PREFIX IN ASCENDING ORDER:
# 'meth', 'eth', 'prop', 'but', 'pent', 'hex', 'hept', 'oct', 'non', 'dec'

formula = gets.chomp

# Write an answer using puts
# To debug: STDERR.puts "Debug messages..."

class Daterminor
  def initialize(formula)
    @formula = formula

    @c_count = count_atom 'C'
    @h_count = count_atom 'H'
    @o_count = count_atom 'O'
  end

  def chemical_name
    prefix = prefix_name
    suffix = suffix_name

    prefix && suffix && c_count_is_valid? ? "#{prefix}#{suffix}" : 'OTHERS'
  end

  private

  def prefix_name
    prefix_names = [nil, 'meth', 'eth', 'prop', 'but', 'pent', 'hex', 'hept', 'oct', 'non', 'dec']

    prefix_names[@c_count]
  end

  def suffix_name
    check_method_and_name = {
      alkane?: 'ane',
      alkene?: 'ene',
      alcohol?: 'anol',
      carboxylic?: 'anoic acid',
      aldehide?: 'anal',
      ketone?: 'anone'
    }

    check_method_and_name.each do |method_name, suffix|
      return suffix if method(method_name).call
    end

    nil
  end

  def alkane?
    @h_count == @c_count * 2 + 2 && @o_count.zero?
  end

  def alkene?
    @h_count == @c_count * 2 && @o_count.zero?
  end

  def alcohol?
    @formula.end_with?('OH') && @h_count == @c_count * 2 + 2
  end

  def carboxylic?
    @formula.end_with?('COOH') && @h_count == @c_count * 2
  end

  def aldehide?
    @formula.end_with?('CHO') && @h_count == @c_count * 2
  end

  def ketone?
    @formula.scan('CO').one? && !@formula.match?(/^CO|CO$/) && @h_count == @c_count * 2
  end

  def c_count_is_valid?
    @o_count <= @formula.match(/COOH|OH|CHO|CO/).to_s.count('O')
  end

  def count_atom(atom)
    @formula.scan(Regexp.new("#{atom}\\d*")).sum do |atom_match|
      [atom_match.delete_prefix(atom).to_i, 1].max
    end
  end
end

daterminor = Daterminor.new formula
result = daterminor.chemical_name

# puts "organic compound name"
puts result

File: python3/practice/classic_puzzle/easy/organic-compound.py
# import sys
# import math

import re

# PREFIX IN ASCENDING ORDER:
# 'meth', 'eth', 'prop', 'but', 'pent', 'hex', 'hept', 'oct', 'non', 'dec'

formula = input()

# Write an answer using print
# To debug: print("Debug messages...", file=sys.stderr, flush=True)

class Daterminor:
    def __init__(self, formula):
        self.formula = formula

        self.c_count = self.__count_atom('C')
        self.h_count = self.__count_atom('H')
        self.o_count = self.__count_atom('O')

    def chemical_name(self):
        prefix = self.__prefix_name()
        suffix = self.__suffix_name()

        return (
            f'{prefix}{suffix}'
            if all([prefix is not None, suffix is not None, self.__is_valid_c_count()])
            else 'OTHERS'
        )

    def __prefix_name(self):
        prefix_names = [None, 'meth', 'eth', 'prop', 'but', 'pent', 'hex', 'hept', 'oct', 'non', 'dec']

        return prefix_names[self.c_count]

    def __suffix_name(self):
        check_method_and_names = {
            self.__is_alkane: 'ane',
            self.__is_alkene: 'ene',
            self.__is_alcohol: 'anol',
            self.__is_carboxylic: 'anoic acid',
            self.__is_aldehide: 'anal',
            self.__is_ketone: 'anone',
        }

        for method, suffix in check_method_and_names.items():
            if method():
                return suffix

        return None

    def __is_alkane(self):
        return self.h_count == self.c_count * 2 + 2 and self.o_count == 0

    def __is_alkene(self):
        return self.h_count == self.c_count * 2 and self.o_count == 0

    def __is_alcohol(self):
        return self.formula.endswith('OH') and self.h_count == self.c_count * 2 + 2

    def __is_carboxylic(self):
        return self.formula.endswith('COOH') and self.h_count == self.c_count * 2

    def __is_aldehide(self):
        return self.formula.endswith('CHO') and self.h_count == self.c_count * 2

    def __is_ketone(self):
        return (
            len(re.findall('CO', self.formula)) == 1
            and not re.match(r'^CO|CO$', self.formula)
            and self.h_count == self.c_count * 2
        )

    def __is_valid_c_count(self):
        return self.o_count <= ''.join(re.findall(r'COOH|OH|CHO|CO', self.formula)).count('O')

    def __count_atom(self, atom):
        return sum(
            max(int(atom_match.removeprefix(atom) or '0'), 1)
            for atom_match
            in re.findall(fr'{atom}\d*', self.formula)
        )

daterminor = Daterminor(formula)
result = daterminor.chemical_name()

# print("organic compound name")
print(result)
